lib/clientv3/networkset.go

```diff
@@ -0,0 +1,101 @@
+// Copyright (c) 2018-2019 Tigera, Inc. All rights reserved.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package clientv3
+
+import (
+	"context"
+
+	apiv3 "github.com/projectcalico/libcalico-go/lib/apis/v3"
+	"github.com/projectcalico/libcalico-go/lib/options"
+	validator "github.com/projectcalico/libcalico-go/lib/validator/v3"
+	"github.com/projectcalico/libcalico-go/lib/watch"
+)
+
+// NetworkSetInterface has methods to work with NetworkSet resources.
+type NetworkSetInterface interface {
+	Create(ctx context.Context, res *apiv3.NetworkSet, opts options.SetOptions) (*apiv3.NetworkSet, error)
+	Update(ctx context.Context, res *apiv3.NetworkSet, opts options.SetOptions) (*apiv3.NetworkSet, error)
+	Delete(ctx context.Context, namespace, name string, opts options.DeleteOptions) (*apiv3.NetworkSet, error)
+	Get(ctx context.Context, namespace, name string, opts options.GetOptions) (*apiv3.NetworkSet, error)
+	List(ctx context.Context, opts options.ListOptions) (*apiv3.NetworkSetList, error)
+	Watch(ctx context.Context, opts options.ListOptions) (watch.Interface, error)
+}
+
+// networkSets implements NetworkSetInterface
+type networkSets struct {
+	client client
+}
+
+// Create takes the representation of a NetworkSet and creates it.  Returns the stored
+// representation of the NetworkSet, and an error, if there is any.
+func (r networkSets) Create(ctx context.Context, res *apiv3.NetworkSet, opts options.SetOptions) (*apiv3.NetworkSet, error) {
```

Comment:
I think we'll want to add some logic like we do for workload endpoints: https://github.com/projectcalico/libcalico-go/blob/master/lib/clientv3/workloadendpoint.go#L58

This ensures that we always set the `projectcalico.org/namespace` label, allowing it to be selected by NetworkPolicy objects.

Note this logic, which enforces that NetworkPolicy selectors are always limited to a particular namespace by adding an additional selector clause: https://github.com/projectcalico/libcalico-go/blob/master/lib/backend/syncersv1/updateprocessors/networkpolicyprocessor.go#L52-L59

We do a similar, though substantially more complicated operation for the rule selectors as well to ensure they are namespaced correctly: https://github.com/projectcalico/libcalico-go/blob/master/lib/backend/syncersv1/updateprocessors/rules.go#L108-L127